src/internal/operators/debounceTime.ts

```diff
@@ -64,74 +65,42 @@ import { lift } from '../util/lift';
-  }
-}
-
-/**
- * We need this JSDoc comment for affecting ESDoc.
- * @ignore
- * @extends {Ignored}
- */
-class DebounceTimeSubscriber<T> extends Subscriber<T> {
-  private debouncedSubscription: Subscription | null = null;
-  private lastValue: T | null = null;
-  private hasValue: boolean = false;
-
-  constructor(destination: Subscriber<T>,
-              private dueTime: number,
-              private scheduler: SchedulerLike) {
-    super(destination);
-  }
-
-  protected _next(value: T) {
-    this.clearDebounce();
-    this.lastValue = value;
-    this.hasValue = true;
-    this.add(this.debouncedSubscription = this.scheduler.schedule(dispatchNext as any, this.dueTime, this));
-  }
-
-  protected _complete() {
-    this.debouncedNext();
-    this.destination.complete();
-  }
-
-  debouncedNext(): void {
-    this.clearDebounce();
-
-    if (this.hasValue) {
-      const { lastValue } = this;
-      // This must be done *before* passing the value
-      // along to the destination because it's possible for
-      // the value to synchronously re-enter this operator
-      // recursively when scheduled with things like
-      // VirtualScheduler/TestScheduler.
-      this.lastValue = null;
-      this.hasValue = false;
-      this.destination.next(lastValue);
-    }
-  }
-
-  private clearDebounce(): void {
-    const debouncedSubscription = this.debouncedSubscription;
-
-    if (debouncedSubscription !== null) {
-      this.remove(debouncedSubscription);
-      debouncedSubscription.unsubscribe();
-      this.debouncedSubscription = null;
-    }
-  }
-}
+export function debounceTime<T>(dueTime: number, scheduler: SchedulerLike = asyncScheduler): MonoTypeOperatorFunction<T> {
+  return (source: Observable<T>) =>
+    lift(source, function (this: Subscriber<T>, source: Observable<T>) {
+      const subscriber = this;
+      let hasValue = false;
+      let lastValue: T | null = null;
+      let debounceSubscription: Subscription | null = null;
 
-function dispatchNext(subscriber: DebounceTimeSubscriber<any>) {
-  subscriber.debouncedNext();
+      source.subscribe(
+        new OperatorSubscriber(
+          subscriber,
+          (value) => {
+            debounceSubscription?.unsubscribe();
+            hasValue = true;
+            lastValue = value;
+            subscriber.add(
+              (debounceSubscription = scheduler.schedule(() => {
+                debounceSubscription = null;
+                if (hasValue) {
+                  hasValue = false;
+                  const value = lastValue!;
+                  lastValue = null;
+                  subscriber.next(value);
+                }
+              }, dueTime))
+            );
+          },
+          undefined,
+          () => {
+            if (hasValue) {
+              subscriber.next(lastValue!);
+              lastValue = null;
+            }
+            subscriber.complete();
+          }
+        )
```

Comment:
question: In `debounce`, AFAICT, a teardown is used to null out the `lastValue` upon explicit unsubscription. Should that be done here, too?

Reply:
I've unified the approach to emitting here, and added more comments.